File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxDiff = 0;
        int high = 0;
        int low = 0;
        
        if (prices.length <= 1)
            return 0;
        
        high = prices[0];
        low = prices[0];
        
        for (int i = 1; i < prices.length; i++){
            // buy low sell high
            low = Math.min(low, prices[i]);
            int curDiff = prices[i] - low;
            maxDiff = Math.max(maxDiff, curDiff);
                
        }
        return maxDiff;
    }
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices or len(prices) == 1:
            return 0
        
        maxSum = 0
        minNum = None
        
        for i in prices:
            if minNum is None:
                minNum = prices[0]
            else:
                minNum = min(minNum, i)
            print(i - minNum)
            maxSum = max(maxSum, i - minNum)
        
        return maxSum
                